doc/tutorial/network_ring.rst

```diff
@@ -0,0 +1,337 @@
+
+.. code-block:: python
+
+   # (5) Create a recipe that generates a network of connected cells.
+   class ring_recipe (arbor.recipe):
+      def __init__(self, ncells):
+         # The base C++ class constructor must be called first, to ensure that
+         # all memory in the C++ class is initialized correctly.
+         arbor.recipe.__init__(self)
+         self.ncells = ncells
+         self.props = arbor.neuron_cable_properties()
+         self.cat = arbor.default_catalogue()
+         self.props.register(self.cat)
+
+      # (6) The num_cells method that returns the total number of cells in the model
+      # must be implemented.
+      def num_cells(self):
+         return self.ncells
+
+      # (7) The cell_description method returns a cell
+      def cell_description(self, gid):
+         return make_cable_cell(gid)
+
+      # The kind method returns the type of cell with gid.
+      # Note: this must agree with the type returned by cell_description.
+      def cell_kind(self, gid):
+         return arbor.cell_kind.cable
+
+      # (8) Make a ring network
+      def connections_on(self, gid):
+         src = (gid-1)%self.ncells
+         w = 0.01
+         d = 5
+         return [arbor.connection(arbor.cell_member(src,0), arbor.cell_member(gid,0), w, d)]
+
+      def num_targets(self, gid):
+         return 1
+
+      def num_sources(self, gid):
+         return 1
+
+      # (9) Attach a generator to the first cell in the ring.
+      def event_generators(self, gid):
+         if gid==0:
+               sched = arbor.explicit_schedule([1])
+               return [arbor.event_generator(arbor.cell_member(0,0), 0.1, sched)]
+         return []
+
+      def probes(self, gid):
+         return [arbor.cable_probe_membrane_voltage('"root"')]
+
+      def global_properties(self, kind):
+         return self.props
+
+The execution context
+*********************
+
+The execution context contains all system-specific information needed by the simulation: it contains the
+thread pool which handles multi-threaded optimization on the CPU; it knows the relevant GPU attributes
+if a GPU is available; and it holds the MPI communicator for distributed simulations.
+The details of the execution context can be customized by the user. However,
+the ideal settings can usually be inferred from the system.
+
+In step **(10)** we specify the number of threads in the thread pool and
+let Arbor know that we have no preference for a particular GPU.
+
+Printing out the :class:`arbor.context` object will show some information of the
+capabilities of the system and the way it was configured. Most Arbor objects can
+be printed out to get obtain some information about its configuration.
+
+.. code-block:: python
+
+   # (10) Set up the hardware context
+   context = arbor.context(threads=12, gpu_id=None)
+   print(context)
+
+The meter manager
```

Comment:
Maybe the meter manager is a distraction? If this is the first time a user is learning about recipes, then it might be best to not introduce any distractions from that core mission.

We can include discussion about meters to a later example, that takes this one, and extends it with performance measurement. Don't be afraid to modify the original example.

Originally, there was a "tutorial", that was separate from the examples, and was going to be extended step by step guide users through the process of going from a single cell to a network model running with MPI.

And I think the approach of having a clear path, that starts simple, and builds up that way with layer on layer is the way to go.

Reply:
I did remove the MPI stuff from this example already. I did not find the meter a distraction. I agree that we should keep things orthogonal-ish, but showing a hint of a feature that will be fully described in a future example/tutorial if it is easy enough to understand is OK with me. The link to the class doc is there (not sure if it works already, but if not it's coming up!). The Arbor meter manager is simple to use, and checkpointing to record runtime and memory consumption is something anyone will understand I think.

Are you saying that I should split this out now? So, have a tutorial/example with just the meter added to `network_ring.rst`?